File: 689. Maximum Sum of 3 Non-Overlapping Subarrays/brute_force/main.py
from typing import List


class Solution:
    def maxSumOfThreeSubarrays(self, nums: List[int], k: int) -> List[int]:
        max_sum, res = 0, []

        def backtrack(i, choosen):
            if len(choosen) == 3:
                nonlocal max_sum, res
                cur_sum = 0
                for j in choosen:
                    cur_sum += sum(nums[j:j+k])
                if cur_sum > max_sum:
                    max_sum = cur_sum
                    res = choosen[:]
                return

            if i >= len(nums) - 1:
                return    
            
            # Choose i
            choosen.append(i)
            backtrack(i + k, choosen)
            choosen.pop()

            # Skip i
            backtrack(i + 1, choosen)
        
        backtrack(0, [])

        return res

File: 689. Maximum Sum of 3 Non-Overlapping Subarrays/brute_force/main.go
package main

func maxSumOfThreeSubarrays(nums []int, k int) []int {
	max_sum, res := 0, []int{}

	var backtrack func(i int, chosen []int)
	backtrack = func(i int, chosen []int) {
		if len(chosen) == 3 {
			cur_sum := 0
			for _, j := range chosen {
				for t := j; t < j+k; t++ {
					cur_sum += nums[t]
				}
			}
			if cur_sum > max_sum {
				max_sum = cur_sum
				res = append([]int{}, chosen...)
			}
			return
		}

		if i > len(nums)-k {
			return
		}

		// chose i
		chosen = append(chosen, i)
		backtrack(i+k, chosen)
		chosen = chosen[:len(chosen)-1]

		// skip i
		backtrack(i+1, chosen)
	}

	backtrack(0, []int{})

	return res
}
